apps/opencs/model/tools/gmstcheck.cpp

```diff
@@ -0,0 +1,91 @@
+#include "gmstcheck.hpp"
+
+#include "../world/defaultgmsts.hpp"
+
+CSMTools::GMSTCheckStage::GMSTCheckStage(const CSMWorld::IdCollection<ESM::GameSetting>& gameSettings)
+    : mGameSettings(gameSettings)
+{}
+
+int CSMTools::GMSTCheckStage::setup()
+{
+    return mGameSettings.getSize();
+}
+
+void CSMTools::GMSTCheckStage::perform(int stage, CSMDoc::Messages& messages)
+{
+    const CSMWorld::Record<ESM::GameSetting>& record = mGameSettings.getRecord (stage);
+    
+    if (record.isDeleted())
+        return;
+    
+    const ESM::GameSetting& gmst = record.get();
+    
+    CSMWorld::UniversalId id (CSMWorld::UniversalId::Type_Gmst, gmst.mId);
+    
+    // Test for empty string
+    if (gmst.mValue.getType() == ESM::VT_String && gmst.mValue.getString().empty())
+        messages.add(id, gmst.mId + " is an empty string", "", CSMDoc::Message::Severity_Warning);
+    
+    // Checking type and limits
+    // optimization - compare it to lists based on naming convention (f-float,i-int,s-string)
+    if (gmst.mId[0] == 'f')
+    {
+        for (size_t i = 0; i < CSMWorld::DefaultGMSTs::FloatCount; ++i)
+        {
+            if (gmst.mId == CSMWorld::DefaultGMSTs::Floats[i])
+            {
+                if (gmst.mValue.getType() != ESM::VT_Float)
+                    messages.add(id, "The type of " + gmst.mId + " is incorrect", 
+                                 "Change the GMST type to a float", CSMDoc::Message::Severity_Error);
```

Comment:
Still the issue with the advice on how to fix it. We don't do that. I suggest to completely remove the hint. We don't need it. The error message should specify the required type though, e.g. The type of xxx is incorrect. It should be a float. Or we could be even more explicit. "Expected float type for xxx but found int type".